### PR TITLE
Laboratory work #1, Buntova Anastasia - 20FPL1

### PR DESCRIPTION
**Pull request by DangerousCrocodile**

### CODE REVIEW

**Comment by ValeriyaKuznetsova** [comment]
It's time to push some changes. October 1st is coming

**Comment by ValeriyaKuznetsova** [comment]
still nothing :(

File: lab_1/main.py [hunk 1]
```diff
@@ -11,7 +12,15 @@ def tokenize(text: str) -> list or None:
     :param text: a text
     :return: a list of lower-cased tokens without punctuation
     """
-    pass
+    if not isinstance(text, str):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:15]: thanks for checks with None at the beginning

File: lab_1/main.py [hunk 2]
```diff
@@ -11,7 +12,15 @@ def tokenize(text: str) -> list or None:
...
+        return None
+    text_without_symbols = ""
+    for i in text:
+        if i == " " or i.isalpha():
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:19]: what if a line is `'I\ngirl'`?

File: lab_1/main.py [hunk 3]
```diff
@@ -11,7 +12,15 @@ def tokenize(text: str) -> list or None:
...
+            text_without_symbols += i
+    good_text = text_without_symbols.lower()
+    tokens = good_text.split()
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:22]: can you split and lower at the same time?

File: lab_1/main.py [hunk 4]
```diff
@@ -21,7 +30,16 @@ def remove_stop_words(tokens: list, stop_words: list) -> list or None:
     :param stop_words: a list of stop words
     :return: a list of tokens without stop words
     """
-    pass
+    if not isinstance(tokens, list):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:33]: join checks

File: lab_1/main.py [hunk 5]
```diff
@@ -21,7 +30,16 @@ def remove_stop_words(tokens: list, stop_words: list) -> list or None:
...
+        return None
+    if not isinstance(stop_words, list):
+        return None
+    blank_list = []
+    for i in tokens:
+        if i not in stop_words:
+            blank_list.append(i)
+    tokens = blank_list
+    return tokens
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:42]: return blank_list just here

File: lab_1/main.py [hunk 6]
```diff
@@ -30,7 +48,14 @@ def calculate_frequencies(tokens: list) -> dict or None:
...
     :return: a dictionary with frequencies
     """
-    pass
+    if not isinstance(tokens, list):
+        return None
+    freq_dict = {}
+    for i in tokens:
+        if isinstance(i, str):
+            some_dict = {i: tokens.count(i)}
+            freq_dict.update(some_dict)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:57]: maybe the previous line in brackets?

File: lab_1/main.py [hunk 7]
```diff
@@ -30,7 +48,14 @@ def calculate_frequencies(tokens: list) -> dict or None:
...
+    return freq_dict or None
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:58]: why `or None`?

File: lab_1/main.py [hunk 8]
```diff
@@ -40,7 +65,20 @@ def get_top_n_words(freq_dict: dict, top_n: int) -> list or None:
     :param top_n: a number of the most common words
     :return: a list of the most common words
     """
-    pass
+    if not isinstance(freq_dict, dict):
+        return None
+    if not isinstance(top_n, int):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:70]: join

File: lab_1/main.py [hunk 9]
```diff
@@ -40,7 +65,20 @@ def get_top_n_words(freq_dict: dict, top_n: int) -> list or None:
...
+        return None
+    inf = list(freq_dict.items())
+    inf = sorted(inf, key=lambda x: x[1], reverse=True)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:73]: sort by keys, not items

File: lab_1/main.py [hunk 10]
```diff
@@ -51,7 +89,15 @@ def create_language_profile(language: str, text: str, stop_words: list) -> dict
     :param stop_words: a list of stop words
     :return: a dictionary with three keys – name, freq, n_words
     """
-    pass
+    if not isinstance(language, str):
+        return None
+    if not isinstance(text, str):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:94]: join

File: lab_1/main.py [hunk 11]
```diff
@@ -62,10 +108,25 @@ def compare_profiles(unknown_profile: dict, profile_to_compare: dict, top_n: int
     :param top_n: a number of the most common words
     :return: the distance
     """
-    pass
+    if not isinstance(unknown_profile, dict):
+        return None
+    if not isinstance(profile_to_compare, dict):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:113]: join

File: lab_1/main.py [hunk 12]
```diff
@@ -74,10 +135,30 @@ def detect_language(unknown_profile: dict, profile_1: dict, profile_2: dict, top
     :param top_n: a number of the most common words
     :return: a language
     """
-    pass
+    if not isinstance(unknown_profile, dict):
+        return None
+    if not isinstance(profile_1, dict):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:140]: check

File: lab_1/main.py [hunk 13]
```diff
@@ -74,10 +135,30 @@ def detect_language(unknown_profile: dict, profile_1: dict, profile_2: dict, top
...
+        return None
+    if not isinstance(profile_2, dict):
+        return None
+    if not isinstance(top_n, int):
+        return None
+    compare_with_1 = compare_profiles(unknown_profile, profile_1, top_n)
+    compare_with_2 = compare_profiles(unknown_profile, profile_2, top_n)
+    lang = ""
+    if compare_with_1 > compare_with_2:
+        lang += profile_1["name"]
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:150]: I'm not sure why you need to add a language to an empty line

File: lab_1/main.py [hunk 14]
```diff
@@ -86,10 +167,44 @@ def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, t
     :return: a dictionary with 7 keys – name, score, common, sorted_common, max_length_word,
     min_length_word, average_token_length
     """
-    pass
+    if not isinstance(unknown_profile, dict):
+        return None
+    if not isinstance(profile_to_compare, dict):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:172]: check

File: lab_1/main.py [hunk 15]
```diff
@@ -86,10 +167,44 @@ def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, t
...
+    for i in popular_words:
+        if i in popular_words_un:
+            top_common.append(i)
+
+    score = len(top_common) / len(popular_words_un)
+
+    new_keys = list((profile_to_compare["freq"]).keys())
+    len_of_words_dict = {}
+    for i in new_keys:
+        len_of_words_dict[len(i)] = i
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:190]: I can see a problem here

if there are tokens with the same length?
wouldn't keys change?

File: lab_1/main.py [hunk 16]
```diff
@@ -98,7 +213,33 @@ def detect_language_advanced(unknown_profile: dict, profiles: list, languages: l
     :param top_n: a number of the most common words
     :return: a language
     """
-    pass
+    if not isinstance(unknown_profile, dict):
+        return None
+    if not isinstance(profiles, list):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:218]: checks

File: lab_1/main.py [hunk 17]
```diff
@@ -98,7 +213,33 @@ def detect_language_advanced(unknown_profile: dict, profiles: list, languages: l
...
+    for i in profiles:
+        if i["name"] in languages:
+            actual_profiles.append(i)
+
+    results = []
+    for profile in actual_profiles:
+        results.append(compare_profiles_advanced(unknown_profile, profile, top_n))
+
+    results = sorted(results, key=lambda x: x["name"])
+    results = sorted(results, key=lambda x: x["score"], reverse=True)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:237]: can you explain what happens here and why you need that?

File: lab_1/main.py [hunk 18]
```diff
@@ -98,7 +213,33 @@ def detect_language_advanced(unknown_profile: dict, profiles: list, languages: l
...
+    if len(results) >= 1:
+        exact_language = (results[0])["name"]
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:239]: why brackets ()?

File: lab_1/main.py [hunk 19]
```diff
@@ -107,7 +248,14 @@ def load_profile(path_to_file: str) -> dict or None:
     :param path_to_file: a path
     :return: a dictionary with three keys – name, freq, n_words
     """
-    pass
+    if not isinstance(path_to_file, str):
+        return None
+    try:
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:253]: why try/except? what are they for?

File: lab_1/main.py [hunk 20]
```diff
@@ -116,4 +264,15 @@ def save_profile(profile: dict) -> int:
     :param profile: a dictionary
     :return: 0 if everything is ok, 1 if not
     """
-    pass
+    if not isinstance(profile, dict):
+        return 1
+    if "name" in profile.keys() is False:
+        return 1
+    if "freq" in profile.keys() is False:
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:271]: join checks

File: lab_1/start.py [hunk 21]
```diff
@@ -19,6 +19,21 @@
...
+        result = main.detect_language_advanced(new_profile, [en, de, la], [], 5)
+        print(result)
+        main.save_profile(new_profile)
+
+    with open(os.path.join(PATH_TO_TEXTS_FOLDER, "la.txt"), "r", encoding="utf-8") as file_to_read:
+        la_text = file_to_read.read()
+
     EXPECTED = 'en'
     RESULT = ''
     # DO NOT REMOVE NEXT LINE - KEEP IT INTENTIONALLY LAST
```
Comment (ValeriyaKuznetsova) [inline lab_1/start.py:39]: assert RESULT == EXPECTED
save smh in result; otherwise, there is no point in the check

File: lab_1/start.py [hunk 22]
```diff
@@ -19,6 +19,21 @@
...
 
+        import main
+
+        en = main.load_profile(os.path.join(PATH_TO_LAB_FOLDER, r"profiles\en.json"))
+        de = main.load_profile(os.path.join(PATH_TO_LAB_FOLDER, r"profiles\de.json"))
+        la = main.load_profile(os.path.join(PATH_TO_LAB_FOLDER, r"profiles\la.json"))
+        new_profile = main.create_language_profile("unk", unknown_text, [])
+        interesting = main.compare_profiles_advanced(new_profile, en, 5)
+        print(interesting)
+        result = main.detect_language_advanced(new_profile, [en, de, la], [], 5)
```
Comment (ValeriyaKuznetsova) [inline lab_1/start.py:30]: can you explain what happens in `start.py`?

File: lab_1/unk.json [hunk 23]
```diff
@@ -0,0 +1 @@
...ht": 1, "the": 15, "bees": 3, "were": 2, "responding": 2, "only": 2, "to": 4, "scent": 1, "of": 5, "foodbut": 1, "what": 1, "did": 2, "third": 1, "dance": 2, "mean": 1, "and": 3, "if": 2, "scenthow": 1, "could": 1, "they": 1, "also": 2, "sniff": 1, "down": 1, "food": 1, "hundreds": 1, "metres": 4, "away": 4, "from": 1, "hive": 1, "foodwhich": 1, "was": 4, "sometimes": 1, "downwind": 1, "on": 1, "a": 2, "hunch": 1, "he": 5, "started": 2, "gradually": 1, "moving": 1, "thefeeding": 1, "dish": 4, "further": 2, "noticed": 1, "as": 1, "so": 1, "that": 1, "dances": 1, "thereturning": 1, "scout": 1, "changing": 1, "placed": 1, "feeding": 2, "over": 1, "ninemetres": 1, "second": 1, "type": 1, "sickle": 1, "version": 1, "came": 1, "into": 1, "playbut": 1, "once": 1, "moved": 2, "it": 1, "past": 1, "scouts": 1, "would": 1, "then": 1, "start": 1, "dancing": 1, "thirdquite": 1, "different": 1, "waggle": 1, "dancethe": 1, "measurement": 1, "actual": 1, "distance": 1, "too": 1, "concluded": 1, "precise": 1, "forexample": 1, "indicated": 1, "by": 1, "complete": 1, "runsthrough": 1, "pattern": 1, "in": 1, "seconds": 1, "when": 1, "awaythe": 1, "number": 1, "dropped": 1, "eleven": 1}, "n_words": 98}
```
Comment (ValeriyaKuznetsova) [inline lab_1/unk.json:1]: remove this file

**Review by ValeriyaKuznetsova — COMMENTED** [review]
Nice one! 

I have some comments on the style and implementation. Also, there are some questions. It's obligatory to answer them here or via Telegram

File: lab_1/main.py (hunk 13 above)
Comment (DangerousCrocodile) [reply lab_1/main.py:150]: Would it be better if there was an empty list instead of empty str?

File: lab_1/main.py (hunk 9 above)
Comment (DangerousCrocodile) [reply lab_1/main.py:73]: Sorry, I don't really understand... all items are sorted by values which are the frequency of words.

File: lab_1/main.py (hunk 15 above)
Comment (DangerousCrocodile) [reply lab_1/main.py:190]: Yes, they probably would. Will it change if I make a dict with words as keys and lengths as values?

File: lab_1/main.py (hunk 17 above)
Comment (DangerousCrocodile) [reply lab_1/main.py:237]: First - dictionaries in list "results" are sorted by names of languages: it means after sorting by score (the next line) dictionaries with the same score will be arranged in alphabetical order.

File: lab_1/main.py (hunk 18 above)
Comment (DangerousCrocodile) [reply lab_1/main.py:239]: OK, i'll remove it)0

File: lab_1/main.py (hunk 19 above)
Comment (DangerousCrocodile) [reply lab_1/main.py:253]: If the path will not be found, it helps to avoid an error and to return None

File: lab_1/start.py (hunk 22 above)
Comment (DangerousCrocodile) [reply lab_1/start.py:30]: I was just checking my code c:
The first line creates a new profile for the unknown text
The second line compares profile made from text in unknown language with english profile in this case
The forth line detects what language it is

File: lab_1/unk.json (hunk 23 above)
Comment (DangerousCrocodile) [reply lab_1/unk.json:1]: But it's needed to compare text in unknown language with other profiles and to detect the language...

File: lab_1/main.py [hunk 24]
```diff
@@ -14,12 +14,14 @@ def tokenize(text: str) -> list or None:
     """
     if not isinstance(text, str):
         return None
+    if "\n" in text:
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:17]: if `\t` is in a text? :)
maybe consider using `isalpha` and `isspace`

File: lab_1/main.py [hunk 25]
```diff
@@ -14,12 +14,14 @@ def tokenize(text: str) -> list or None:
...
+        for i in range(text.count("\n")):
+            text = text.replace("\n", " ")
     text_without_symbols = ""
     for i in text:
         if i == " " or i.isalpha():
             text_without_symbols += i
-    good_text = text_without_symbols.lower()
-    tokens = good_text.split()
+    tokens = (text_without_symbols.lower()).split()
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:24]: brackets aren't needed

File: lab_1/main.py (hunk 9 above)
Comment (ValeriyaKuznetsova) [reply lab_1/main.py:73]: yes, and I tell you to sort keys(),` freq_dict.items()` -> `freq_dict.keys()` or better `freq_dict`
This way you don't need extra `for` loop

File: lab_1/main.py [hunk 26]
```diff
@@ -135,26 +131,21 @@ def detect_language(unknown_profile: dict, profile_1: dict, profile_2: dict, top
...
+        lang.append(profile_1["name"])
     elif compare_with_1 < compare_with_2:
-        lang += profile_2["name"]
+        lang.append(profile_2["name"])
     elif compare_with_1 == compare_with_2:
-        lang += [profile_1["name"], profile_2["name"]].sort()
-    if isinstance(lang, list):
-        lang = lang[0]
-    return lang
+        (lang.append(profile_1["name"], profile_2["name"])).sort()
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:147]: no brackets needed

File: lab_1/main.py (hunk 19 above)
Comment (ValeriyaKuznetsova) [reply lab_1/main.py:242]: but why to use it? I don't get it
A simple `If` will do

**Review by ValeriyaKuznetsova — COMMENTED** [review]
Two pending comments remain. See new minor ones

**Comment by ValeriyaKuznetsova** [comment]
In general, good job! In some places, you overcomplicated things :) 

Correct these issues by Friday

**Comment by ValeriyaKuznetsova** [comment]
And please, don't resolve issues; otherwise, it's hard to check

File: lab_1/main.py (hunk 24 above)
Comment (DangerousCrocodile) [reply lab_1/main.py:17]: Oh, I didn't know about isspace. Thanks, it makes things much easier)0

File: lab_1/main.py (hunk 25 above)
Comment (DangerousCrocodile) [reply lab_1/main.py:24]: Horosho-horosho

File: lab_1/main.py (hunk 26 above)
Comment (DangerousCrocodile) [reply lab_1/main.py:147]: Understood

**Pull request closed without merging**